Add Json Schema Validation for publishednodes.json file

WilliamBerryiii opened the pull request:
# Description
This PR begins the process of adding JSON-Schema based validation for the Azure OPC Publisher's [`publishednodes.json`](https://raw.githubusercontent.com/Azure/iot-edge-opc-publisher/main/opcpublisher/publishednodes.json) configuration file. A schema generation tool, [the OPC Publisher Schema Validator](https://github.com/WilliamBerryiii/opcpublisherschemavalidator), was created to programmatically generate a json-schema file that adheres to the three (3) supported (documented) formats for OPC UA node configuration within the [`publishednodes.json`](https://raw.githubusercontent.com/Azure/iot-edge-opc-publisher/main/opcpublisher/publishednodes.json) file.   The "default" generated schema, included in this PR, enforces not only proper object construction within the configuration file, but also validates property values such as NodeId, EndpointUrls, etc. (further details below)

Fixes #1301 
Partially addressed #1032 

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

## How Has This Been Tested?

This PR includes significant changes to existing OPC Publisher test code, including additions for 'schema validation error' validation, and ensuring that all pre-existing tests were/are exercised in accordance with their original conditional intent.

Further, the schema generator too covers both positive and negative case testing for the schema generation process as well as the output schema file (subsequently included in this PR).  Generated schema tests in the tooling can be found in the [generator tests](https://github.com/WilliamBerryiii/opcpublisherschemavalidator/tree/main/test).

## Contributors

@WilliamBerryiii - Developer of [OPC Publisher Schema Generator/Validator](https://github.com/WilliamBerryiii/opcpublisherschemavalidator) tool
@suneetnangia - Implemented all code changes within this PR

## Dependencies

This contribution adds an additional package dependency:

[JsonSchema.Net](https://github.com/gregsdennis/json-everything) @ Version 1.10.7

## Checklist

- [X] We have performed a self/peer-review of contributed code
- [X] We have commented code, particularly in hard-to-understand areas and updated method parameter descriptions
- [X] We have made corresponding changes to the documentation
- [X] Our changes generate no new warnings
- [X] We have added tests that prove the change is effective and works
- [X] New and existing unit tests pass locally with all changes

## Notes for Reviewers/Maintainers

There are a few points of consideration for this PR that would be useful to discuss:

- *The new feature is controlled through a CLI flag* We have made the schema validation check feature toggleable through a CLI flag so as to not impact existing implementations where this feature may not be wanted/needed.
- *The schema generation tool is external to this repository.* The tool we used to programmatically generate the schema file is external to this repository. It may be of use to include it as part of this package - perhaps we can discuss in another issue/thread.  The tool can be found [here](https://github.com/WilliamBerryiii/opcpublisherschemavalidator) and it includes a number of design decision points, particularly around property value validation, that merit review. 
- *Further expand unit test coverage.* There are a number of added unit tests covering functionality, though it may be of the reviewer's interest to have further coverage/edge case protections put in place. We are happy to take feedback on this an update accordingly.

suneetnangia committed:
Added Json schema validation of publishednodes.json, with backward compatibility and jobs deserialization bug fix.

suneetnangia committed:
Updated with latest json schema version and removed any additional properties added previously in tests to conform to previous schema versions.

marcschier commented on common/src/Microsoft.Azure.IIoT.Serializers.NewtonSoft/src/Microsoft.Azure.IIoT.Serializers.NewtonSoft.csproj:
```diff
@@ -8,6 +8,7 @@
   </PropertyGroup>
   <ItemGroup>
     <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
+    <PackageReference Include="Newtonsoft.Json.Schema" Version="3.0.14" />
```
this library has an incompatible license (commercial license or AGPL).  Is it possible to use a different library here as we would not be able to get this through our governance.

WilliamBerryiii replied:
We will absolutely investigate and update. Thank you for catching that!

WilliamBerryiii replied:
@marcschier - we have updated the PR with a replacement for the library with an incompatible license. I have also updated the initial PR description to include notes about the changes we made in the subsequent commit.

WilliamBerryiii committed:
Initial implementation of json-everything as replacement for Newtonsoft Json Schema (#2)

Replace Newtonsoft Schema validation with JsonSchemaDotNet
Co-authored-by: Suneet Nangia <suneetnangia@gmail.com>

WilliamBerryiii committed:
remove lingering references to Newtonsoft.Json.Schema library

WilliamBerryiii committed:
remove duplicate schema file and convert to linked asset with copy always for tests

WilliamBerryiii committed:
Delete modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/schemas directory

GitHub seems to think this folder still exists though it is not in the repository anymore. Removing in an attempt to clean up the PR.

WilliamBerryiii committed:
Revert version change to sln file, address LGTM automated review feedback.

WilliamBerryiii commented:
> This pull request **introduces 1 alert** when merging [9c1a153](https://github.com/Azure/Industrial-IoT/commit/9c1a153efb04fe917ddfe9674175fb8c4ac78116) into [53a5d23](https://github.com/Azure/Industrial-IoT/commit/53a5d23d43a9187b9ee7352add6442cad532d4a3) - [view on LGTM.com](https://lgtm.com/projects/g/Azure/Industrial-IoT/rev/pr-e9918afc175afebc980c68b8ea2652f088a8372c)
> 
> **new alerts:**
> 
> * 1 for Useless assignment to local variable

Addressed LGTM automated review issues in commit [47c6f9d](https://github.com/Azure/Industrial-IoT/pull/1305/commits/47c6f9d8af2192afed9928f93258400757809cdf)

marcschier committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

karok2m committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

mregen committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

karok2m committed:
Merge branch 'main' into json_schema_validation

karok2m commented on common/src/Microsoft.Azure.IIoT.Abstractions/src/Validators/JsonSchemaValidationResult.cs:
```diff
@@ -0,0 +1,45 @@
+﻿namespace Microsoft.Azure.IIoT.Validators {
```
Copyright notice is missing.

karok2m commented on common/src/Microsoft.Azure.IIoT.Validators.JsonSchemaDotNet/src/Default/JsonSchemaDotNetSchemaValidator.cs:
```diff
@@ -0,0 +1,71 @@
+﻿namespace Microsoft.Azure.IIoT.Validators {
```
Copyright notice is missing.

karok2m commented on common/src/Microsoft.Azure.IIoT.Serializers.NewtonSoft/src/Microsoft.Azure.IIoT.Serializers.NewtonSoft.csproj:
```diff
@@ -10,6 +10,7 @@
     <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
   </ItemGroup>
   <ItemGroup>
+    <ProjectReference Include="..\..\Microsoft.Azure.IIoT.Validators.JsonSchemaDotNet\src\Microsoft.Azure.IIoT.Validators.JsonSchemaDotNet.csproj" />
```
We do not use `DotNet` in the names of our projects or classes. Please remove this part from the project name.

karok2m commented on common/src/Microsoft.Azure.IIoT.Validators.JsonSchemaDotNet/src/Default/JsonSchemaDotNetSchemaValidator.cs:
```diff
@@ -0,0 +1,71 @@
+    /// <summary>
+    /// JsonSchema.Net schema validator.
+    /// </summary>
+    public class JsonSchemaDotNetSchemaValidator : IJsonSchemaValidator {
```
We do not use `DotNet` in the names of our projects or classes. Please remove this part from the class name.

suneetnangia replied:
JsonSchemaDotNet (JsonSchema.net) is the name of nuget package which we use for schema validation hence it's reflected in the project and class name(s). This follows the existing naming convention of NewtonSoft nuget package where it's reflected in class names like NewtonSoftJsonSerializer.cs. Wonder if we still need to change the names?

karok2m replied:
OK, I see the reasoning behind the naming here. I will mark this and other comment about `DotNet` as resolved.

karok2m commented on common/src/Microsoft.Azure.IIoT.Abstractions/src/Serializers/Extensions/SerializerEx.cs:
```diff
@@ -4,9 +4,9 @@
 
 namespace Microsoft.Azure.IIoT.Serializers {
-    using Microsoft.Azure.IIoT.Http;
+    using Microsoft.Azure.IIoT.Http;    
```
Please remove trailing whitespaces introduced by this PR.

suneetnangia replied:
Removing all trailing whitespaces from the solution, including in this PR. Subsequently, as agreed, raising a separate PR to include styling checks in editorconfig so code styling checks are part of the code for devs when they make changes in future.

suneetnangia committed:
Removed trailing whitespaces from solution by temporarily using SA1028 rule locally + added missing copyright notice.

karok2m commented on components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Engine/publishednodes.json:
```diff
@@ -1,19 +1,25 @@
 [
   {
     "EndpointUrl": "opc.tcp://server1:49580",
+    "UseSecurity": false,
```
Let's keep the test with previous version of `publishednodes.json` and add new test for the new version as all of those fields are optional and can be skipped. This way we can be sure that publisher can still consume older `publishednodes.json` files.

suneetnangia replied:
Adding InlineData in test to check publishednodes.json with previous and new versions both, presuming that should cover what we need?

suneetnangia replied:
InlineData data changes will be pushed in next commit. Please feel free to suggest otherwise.

karok2m commented on components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Engine/pn_assets.json:
```diff
@@ -11,7 +11,9 @@
       {
         "Id": "nsu=http://microsoft.com/Opc/OpcPlc/;s=StepUp",
-        "DataSetFieldId": "nsu=http://microsoft.com/Opc/OpcPlc/;s=StepUp"
+        "DataSetFieldId": "nsu=http://microsoft.com/Opc/OpcPlc/;s=StepUp",
```
Same comment as previously. Let's keep the version of the file without `DataSetFieldId`, `OpcSamplingInterval` and `OpcPublishingInterval` properties and have a new file with those to make sure OPC Publisher would recognize both as having valid schemas.

suneetnangia replied:
Assuming you meant to say "keep DataSetFieldId" in the above comment.

suneetnangia replied:
Adding InlineData in test to check pn_assets.json with previous and new versions both, presuming that should cover what we need?

suneetnangia replied:
InlineData data changes will be pushed in next commit. Please feel free to suggest otherwise.

karok2m replied:
Thanks.

karok2m commented on common/src/Microsoft.Azure.IIoT.Serializers.NewtonSoft/src/Default/NewtonSoftJsonSerializer.cs:
```diff
@@ -67,14 +80,31 @@ public class NewtonSoftJsonSerializer : IJsonSerializerSettingsProvider,
+                    if (validationResults.Any(r => r.IsValid == false)) {
+                        var validationResultsMessage =
+                            string.Join("; ",
+                                validationResults.Select(r =>
```
I think that we need to print messages only from the results that are invalid, meaning we need to first filter on `(r => !r.IsValid)` before joining.

karok2m replied:
Can we please have a test as well that will make sure that only messages from invalid validation results are printed ?

suneetnangia replied:
Thanks, that's a valid point, and this is being checked in test called `EnsureMultipleValidationErrorsAreFlattened` i.e. we throw multiple validation errors in this test and validate that all items in `validationResults` are invalid. So, no valid messages are present in collection when schema validation fails (for single or multiple errors) but I think I agree, having a check here would be more reassuring if the underlying implementation change in future.

suneetnangia replied:
Plus checked in `PnPlcSchemaValidationFailureTest` test from deserializer perspective.

karok2m replied:
@suneetnangia, I wanted to come up with a small example that would have a valid object and then an invalid one to see if the valid one is present in the result or not, so I modified `ConfigurationFileWithLargeErrorListIsHandledSuccessfully()` test and provided it with the following content:

```json
[
  {
    "EndpointUrl": "opc.tcp://20.185.195.172:53530/OPCUA/SimulationServer",
    "UseSecurity": false,
    "OpcNodes": [
      {
        "Id": "i=1001",
        "OpcSamplingInterval": 2000,
        "OpcPublishingInterval": 5000,
      },
      {
        "OpcSamplingInterval": 2000,
        "OpcPublishingInterval": 5000,
      }
    ]
  }
]
```

As you can see the second object in `OpcNodes` list does not have an `Id` property. And to my surprise validation result was that this was a valid content for published nodes file. Returned array contained only one element that had its `IsValid` property equal to `true`. Can you please have a look at this sample and see why is it not reporting an invalid content here and complaining that second element should have an `Id` field ?

suneetnangia replied:
Investigating... on the face of it we probably need to fix this in Json schema (cc @WilliamBerryiii ), added additional test to cover this, thanks for highlighting it.

WilliamBerryiii replied:
I'm working to modify my schema generation tool to improve this issue and will additionally add a battery of tests to cover various conditions once the schema is updated.

suneetnangia replied:
Schema changes will be pushed in next commit.

karok2m replied:
Thanks for fixing this.

karok2m commented on common/src/Microsoft.Azure.IIoT.Serializers.NewtonSoft/src/Default/NewtonSoftJsonSerializer.cs:
```diff
@@ -67,14 +80,31 @@ public class NewtonSoftJsonSerializer : IJsonSerializerSettingsProvider,
+                                validationResults.Select(r =>
+                                    $"Validation failed with error: {r.Message} at schema path: {r.SchemaLocation}, and configuration file location {r.InstanceLocation}"));
+
+                        throw new JsonSerializationException(validationResultsMessage);
```
`JsonSerializationException` object created here will be caught on line `109` bellow and transformed into `SerializerException` object. So I would suggest to throw `SerializerException` here instead.

suneetnangia replied:
Thanks, it was done this way to avoid the breaking change i.e. existing deserialization code would throw `JsonSerializationException` which is wrapped by `SerializerException`. By keeping the same wrapping with inner exception being `JsonSerializationException` we maintain the callers treatment of exceptions from this method.

karok2m replied:
OK, got you. Then let's keep it as it is.

karok2m commented on docs/modules/publisher-commandline.md:
```diff
@@ -3,8 +3,11 @@
-                                       If this Option is specified it puts OPC Publisher into standalone  mode.
+                                       the filename to configure the nodes to publish.
+                                       If this option is specified it puts OPC Publisher into standalone mode.
+             --pfs, --publishfileschema=VALUE
```
Let's create a new heading in this file for options introduced after `2.8.0` version as listing this argument under `for Version 2.6 and above` is misleading.

suneetnangia replied:
Version changes will be pushed in next commit.

karok2m replied:
@suneetnangia, I would still suggest to add a new markdown heading for newly introduced command line arguments like the one bellow. The current changes introduce a gap in between `2.5` and `2.8` which will be confusing for people reading the file and for people still using `2.6` or `2.7` versions.

`OPC Publisher Command Line Arguments Introduced After Version 2.8.0`

WilliamBerryiii replied:
in the next commit, I have put this under the 2.8.0 section. Given that 2.8.0 is an LTS it feels more natural for the additions to come after the base (e.g. LTS).  Let me know if you want it bumped to the top of the file.

karok2m replied:
@WilliamBerryiii, can you please restore the original naming of the first paragraph that used to say `OPC Publisher Command Line Arguments for Version 2.6 and above` instead of `OPC Publisher Command Line Arguments for Version 2.8.0 and above` ? This has been changed in one of the previous commits. Everything else looks great.

WilliamBerryiii replied:
Reverted per commit: https://github.com/Azure/Industrial-IoT/pull/1305/commits/fa262e34b5e40d67d3fe8eb7c1feece3a93e6fb2

Thanks for clarification on that; and really appreciate all the support from you to land this!

karok2m commented on docs/modules/publisher.md:
```diff
@@ -109,8 +109,8 @@ To persist the security configuration of OPC Publisher across restarts, the cert
 
-The simplest way to configure OPC Publisher is via a configuration file. An example configuration file is provided via the file [`publishednodes.json`](publishednodes.json).
-Configuration file syntax has changed over time and OPC Publisher still can read old formats, but converts them into the latest format when persisting the configuration, done regularly in an automated fashion.
+The simplest way to configure OPC Publisher is via a configuration file. An example configuration file is provided via the file [`publishednodes.json`](https://raw.githubusercontent.com/Azure/iot-edge-opc-publisher/main/opcpublisher/publishednodes.json).
```
Let's keep the previous link to just `publishednodes.json` as `iot-edge-opc-publisher` repo is archived and it redirects to `Industrial-IoT` repo.

suneetnangia replied:
Do you still want us to update link to just `publishednodes.json`?

karok2m commented on modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Schemas/publishednodesschema.json:
```diff
@@ -0,0 +1,184 @@
+                "properties": {
+                  "Id": {
+                    "type": "string",
+                    "pattern": "(^nsu=http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*(),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+/;(i=(\\d+)$))|(^nsu=http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*(),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+/;(s=([\\x00-\\x7F]|([\\xC2-\\xDF]|\\xE0[\\xA0-\\xBF]|\\xED[\\x80-\\x9F]|(|[\\xE1-\\xEC]|[\\xEE-\\xEF]|\\xF0[\\x90-\\xBF]|\\xF4[\\x80-\\x8F]|[\\xF1-\\xF3][\\x80-\\xBF])[\\x80-\\xBF])[\\x80 ...
```
Is there a way to add a comment in this JSON file to explain what are the possible value for the `Id` field ? At the moment it is very hard to make any sense of this pattern. Alternatively, is it possible to add a `publishednodesschema.md` file in the same directory with those comments ?

WilliamBerryiii replied:
@karok2m - Absolutely. The schema file is generated by a reasonably well documented tool we built, because of the complexity you noted.  

We can pull over some more of the generator tool tests as well as the documentation if you like, OR, 
we can add the tool to this repo, if you see value in folks being able to generate custom schema for their domains; 

The tool allows you to customize the regular expressions and validation routines; e.g. if your OPC UA servers are only available within a given subdomain or on a specific port, you can modify the validation regular expression with those additional constraints and programmatically generate a new schema file for inclusion in a CI/CD pipeline, as part of DevEx or within OPC Publisher. 

The schema generator can be found under my GH alias here: [https://github.com/WilliamBerryiii/opcpublisherschemavalidator](https://github.com/WilliamBerryiii/opcpublisherschemavalidator) 

Let us know if you have a preference for: 

- just the documentation
- just the tool
- the tool and the documentation

suneetnangia replied:
Documentation changes (comments in schema and supporting md doc) will be pushed in next commit. Please feel free to suggest otherwise.

karok2m replied:
Thanks a lot for adding comments to the schema file and a supporting markdown file. It looks great.

karok2m commented on modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Schemas/publishednodesschema.json:
```diff
@@ -0,0 +1,184 @@
+                "properties": {
+                  "ExpandedNodeId": {
+                    "type": "string",
+                    "pattern": "(^nsu=http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*(),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+/;(i=(\\d+)$))|(^nsu=http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*(),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+/;(s=([\\x00-\\x7F]|([\\xC2-\\xDF]|\\xE0[\\xA0-\\xBF]|\\xED[\\x80-\\x9F]|(|[\\xE1-\\xEC]|[\\xEE-\\xEF]|\\xF0[\\x90-\\xBF]|\\xF4[\\x80-\\x8F]|[\\xF1-\\xF3][\\x80-\\xBF])[\\x80-\\xBF])[\\x80 ...
```
Please add a comment either in this JSON file (if supported) or in a markdown file near it (let's say `publishednodesschema.md`) with explanation or a few samples of valid values for `ExpandedNodeId` field.

WilliamBerryiii replied:
Will address pending response on https://github.com/Azure/Industrial-IoT/pull/1305#discussion_r702944645

karok2m commented on modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Schemas/publishednodesschema.json:
```diff
@@ -0,0 +1,184 @@
+            "properties": {
+              "Identifier": {
+                "type": "string",
+                "pattern": "(^i=(\\d+)$)|(^s=([\\x00-\\x7F]|([\\xC2-\\xDF]|\\xE0[\\xA0-\\xBF]|\\xED[\\x80-\\x9F]|(|[\\xE1-\\xEC]|[\\xEE-\\xEF]|\\xF0[\\x90-\\xBF]|\\xF4[\\x80-\\x8F]|[\\xF1-\\xF3][\\x80-\\xBF])[\\x80-\\xBF])[\\x80-\\xBF])+$)|(^g=([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})$)|(^b=(?:[A-Za-z\\d+/]{4})*(?:[A-Za-z\\d+/]{3}=|[A-Za-z\\d+/]{2}==)$)|(^ns=(\ ...
```
Same comment as for `Id` and `ExpandedNodeId` fields.

WilliamBerryiii replied:
Will address pending response on https://github.com/Azure/Industrial-IoT/pull/1305#discussion_r702944645

karok2m commented on modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Runtime/LegacyCliConfigKeys.cs:
```diff
@@ -27,6 +32,11 @@ public static class LegacyCliConfigKeys {
+        /// <summary>
+        /// Key for the specified published nodes schema filename.
+        /// </summary>
+        public const string PublisherNodeConfigurationSchemaFilename = "PublishedNodesSchemaFile";
```
Please change `PublisherNode` part of the name to `PublishedNodes` to be consistent.

suneetnangia replied:
Thanks, we tried to be consistent with the existing one above it i.e. "PublisherNodeConfigurationFilename", wonder if we should update that to be "PublisherNode**s**ConfigurationFilename" as well?

karok2m replied:
Yes, please change the name of that one as well.

suneetnangia replied:
Name changes will be pushed in next commit.

karok2m replied:
@suneetnangia, I see that you added `s` after `PublisherNode`. Can you also change `Publisher` to `Published` (with `ed` in the end) please ? I now understand now that those two look too similar to each other and I should have called it out in the first comment. If you want to, I can also push this comment, just let me know.

suneetnangia replied:
Np, a fix coming in next commit.

karok2m commented on common/src/Microsoft.Azure.IIoT.Validators.JsonSchemaDotNet/tests/Default/SchemaValidationTests.cs:
```diff
@@ -0,0 +1,207 @@
+            var validator = new JsonSchemaDotNetSchemaValidator();
+            var results = validator.Validate(Encoding.UTF8.GetBytes(alteredConfig), schemaReader);
+
+            // Wnsure that we failed the regex NodeID check
```
`Wnsure` -> `Ensure`

suneetnangia replied:
To be fixed.

suneetnangia replied:
Typo changes will be pushed in next commit.

karok2m commented on common/src/Microsoft.Azure.IIoT.Abstractions/src/Serializers/Extensions/SerializerEx.cs:
```diff
@@ -167,16 +169,34 @@ public static class SerializerEx {
+        /// <param name="reader"></param>
+        /// <param name="schemaReader"></param>
+        /// <returns></returns>
+        public static T Deserialize<T>(this ISerializer serializer,
```
We can also add `TextReader schemaReader = null` parameter to the `Deserialize()` method above on line `167` and remove this extra method. That way all methods accepting `schemaReader` would have `null` default for it. What do you think ?

suneetnangia replied:
IMHO if we were to use default schemaReader which could be overriden optionally by caller of this method then I'd use optional parameter otherwise method overload. One thing which probably makes this distinction more visible is that if we do null checks in these public methods then overload method would throw error is null is provided for schemaReader but with optional we would not know the intention of the caller. Totally your call as a maintainer of code.

suneetnangia replied:
No changes made regarding this in the next commit, please feel free to suggest otherwise.

karok2m replied:
I wanted to have a consistent usage of `TextReader schemaReader` parameter. The other `Deserialize()` calls in the class that have a `schemaReader` have it as an optional parameter with `null` default value. But I am also happy with keeping this as it is as well, it's a minor comment.

karok2m commented on components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs:
```diff
@@ -191,16 +191,27 @@ public class LegacyJobOrchestrator : IJobOrchestrator {
                     var currentFileHash = GetChecksum(_legacyCliModel.PublishedNodesFile);
                     var availableJobs = new ConcurrentQueue<JobProcessingInstructionModel>();
-                    if (currentFileHash != _lastKnownFileHash) {
+                    if (currentFileHash != _lastKnownFileHash || retryCount < 3) {
```
Why do we need additional condition here on `retryCount` ?

suneetnangia replied:
This was something I ran into while testing the changes, if we do not have this check and exception is raised in `if` block at line 194, retries do not work as `currentFileHash != lastKnownFileHash` condition is not true anymore due to `_lastKnownFileHash = currentFileHash;` assignment in the `if` block. To make if block aware of retries are in play, `retryCount < 3` was added. Other way possibly to fix this could be to add `_lastKnownFileHash = string.Empty;` in

```
catch (IOException) {
         throw; //pass it thru, to handle retries
}
```

suneetnangia replied:
No changes made regarding this in the next commit, please feel free to suggest otherwise.

karok2m replied:
OK, got you. Thanks for the fix.

koepalex committed:
Merge branch 'main' into json_schema_validation

koepalex committed:
Merge branch 'main' into json_schema_validation

WilliamBerryiii committed:
Second Item Array Failure Schema Fix

WilliamBerryiii committed:
add schema spec documentation.

WilliamBerryiii committed:
udpate schema with additional comments from code reivew.

WilliamBerryiii committed:
add endpoint url regex design notes to design doc.

suneetnangia committed:
Addressed PR review comments.

karok2m committed:
Merge branch 'main' into json_schema_validation

karok2m commented on modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Schemas/schema_spec.md:
```diff
@@ -0,0 +1,74 @@
+- `s` - string based NodeIds -  (RegEx matches UTF-8 character set)
+
+Each of these NodeId types has a discrete regular expression to ensure a well formed document.
+Note that the `string` type NodeId regular expression encludes all UTF-8 characters, only
```
Typo: `encludes` -> `includes`.

karok2m commented on modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Schemas/publishednodesschema.json:
```diff
@@ -0,0 +1,188 @@
+	"$comment": "The outer most object of the configuration file must be an array, though its contents may adhere to several differing schema, presented from newest to oldest supported schema. Further design details can be found in the markdown document that is adjacent to this schema.",
+	"type": "array",
+	"items": {
+		"$comment": "The nested 'oneOf' under 'items' ensures that each array element must comform to one of the three following subschema. See the following stackoverflow post for details: https://stackoverflow.com/a/67314134/1276028",
```
Typo: `comform` -> `conform`.

karok2m commented on modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Schemas/publishednodesschema.json:
```diff
@@ -0,0 +1,188 @@
+							"properties": {
+								"Id": {
+									"type": "string",
+									"$comment": "This subschema supports all allowabe formats, e.g. `nsu={url/urn resourece path}/;{nodeId type}={nodeId}` or `ns={namespace index based on OPC UA IM}/;{nodeId type}={nodeId}` or `{nodeId type}={nodeId}`",
```
Typos:
* `allowabe` -> `allowable`
* `resourece` -> `resource`

karok2m commented on modules/src/Microsoft.Azure.IIoT.Modules.OpcUa.Publisher/src/Schemas/publishednodesschema.json:
```diff
@@ -0,0 +1,188 @@
+							"properties": {
+								"ExpandedNodeId": {
+									"type": "string",
+									"$comment": "This subschema only supports the use of the expanded nodeid format, e.g. `nsu={url/urn resourece path}/;{nodeId type}={nodeId}`",
```
Typo: `resourece` -> `resource`.

karok2m commented on common/src/Microsoft.Azure.IIoT.Validators.JsonSchemaDotNet/tests/Default/SchemaValidationTests.cs:
```diff
@@ -0,0 +1,224 @@
+        [Fact]
+        // There are nuances to the use of "items" schema for arrays in JSON Schema.
+        // Particualrly the difference between "items: [{ ... }]" and "items: { ... }"
+        // The first checks ONLY the schema of the array element in in a given array index position,
```
Typo: `in` is used twice.

WilliamBerryiii committed:
Correct typos across docs, schema and tests.

WilliamBerryiii committed:
Update command line doc to include post 2.8.0 additional args

WilliamBerryiii committed:
Update Publisher* to Published* in Legacy CLI files.

suneetnangia committed:
Changed casing from this -> This in schema.

WilliamBerryiii committed:
Revert heading changes in publisher cmd line docs 

This change reverts the heading for the heading for the publisher cmd line args doc back to the original version per code upstream review comment.

karok2m committed:
Merge branch 'main' into json_schema_validation

WilliamBerryiii commented:
@karok2m, @marcschier, @koepalex, @hansgschossmann - Thank you very much for the support you've given @suneetnangia and myself on developing and integrating this capability.
